### PR TITLE
Crop Image functionality

### PR DESCRIPTION
**Pull request by vinothvino42**
Ticket: [#627](https://github.com/EnsembleUI/ensemble/issues/627)
In this PR, I added image crop action

Updated with new package - https://pub.dev/packages/custom_image_crop

NEW EDL
```yaml
View:
  header:
    title: Image Cropper
    styles:
      color: white
      backgroundColor: black
  styles:
    scrollableView: false
  body:
    Column:
      styles:
        gap: 24
        padding: 24
        crossAxis: stretch
        backgroundColor: black
      children:
        - ImageCropper:
            id: imageCropperId
            source: https://ensembleui.com/assets/images/hero_social.jpg
            styles:
              {
                height: 350,
                width: 350,
                borderWidth: 1,
                borderRadius: 10,
                fit: fill,
                shape: square,
                cropPercentage: 0.8,
                backgroundColor: 0xFFA9A9A9,
                strokeColor: white,
                strokeWidth: 3,
                isScale: true,
                isMove: true,
                isRotate: true,
              }
            onCropped:
              executeCode:
                body: |
                  //@code
                  console.log("Image Cropped Successfully");
                  console.log(event.data.file);

        - Row:
            styles:
              gap: 24
              padding: 24
              mainAxis: center
            children:
              - Icon:
                  name: refresh
                  styles:
                    size: 30
                    color: white
                  onTap:
                    executeCode:
                      body: |
                        //@code
                        imageCropperId.reset();

              - Icon:
                  name: crop
                  styles:
                    size: 30
                    color: white
                  onTap:
                    executeCode:
                      body: |
                        //@code
                        imageCropperId.crop();

              - Icon:
                  name: rotate_left
                  styles:
                    size: 30
                    color: white
                  onTap:
                    executeCode:
                      body: |
                        //@code
                        imageCropperId.rotateLeft();
              - Icon:
                  name: rotate_right
                  styles:
                    size: 30
                    color: white
                  onTap:
                    executeCode:
                      body: |
                        //@code
                        imageCropperId.rotateRight();

```

Properties
1. shape - circle | square | rectangle
2. strokeColor and strokeWidth

Methods
1. reset()
3. crop()
4. rotateLeft()
5. rotateRight()

### CODE REVIEW

**Comment by vusters** [comment]
@vinothvino42 this is a specialize behavior that doesn't make sense for anything else with Image. Instead of making the user wire things up. Simply have a `CropableImage`.

**Comment by vinothvino42** [comment]
> @vinothvino42 this is a specialize behavior that doesn't make sense for anything else with Image. Instead of making the user wire things up. Simply have a `CropableImage`.

@vusters @amin-nas 
So let's say we're cropping an image with the `CroppableImage` widget.
Then we've to create a new separate ensemble page with croppable and some buttons. Is it a really good user experience?

Why do we need the `CroppableImage` widget? It's used only for cropping and not often too. So why not just have an `cropImage` action that will open the crop page by default with the image file path and get the file object once it's done.

**Comment by vusters** [comment]
cropable image has more properties so you don't want to pollute Image which is very simply. The ImageCropper can handle its own popup - you don't need another screen for it:
```
ImageCropper:
  source:
  overlayWidget:          # in case you want to overlay a button or some notification to tell the user the image is editable
  onCropped:                # call Action to send the updated image to the back end
```

**Comment by vinothvino42** [comment]
@vusters I just updated it. Can you check my new changes

```yaml
View:
  header:
    title: Image Cropper
  styles:
    scrollableView: true
  body:
    Column:
      styles:
        gap: 24
        padding: 24
      children:
        - ImageCropper:
            id: imageCropperId
            source: example.png
            cornerDotColor: red
            maskColor: blue
            cornerRadius: 25
            onStatusChanged:
              executeCode:
                body: |
                  //@code
                  console.log("Status Changed: Status");
                  if (event.data.cropStatus == "ready") {
                    console.log("Image Ready - ");
                    console.log(event.data.path);
                    displayCroppedImageId.source = event.data.path;
                  }
            onCropped:
              executeCode:
                body: |
                  //@code
                  console.log("Image Cropped Successfully");

        - Button:
            label: Crop Image
            onTap:
              executeCode:
                body: |
                  //@code
                  imageCropperId.crop();

        - Image:
            id: displayCroppedImageId
            source: ${event.data.path}

```

**Comment by vusters** [comment]
@vinothvino42 @amin-nas how about the below. There is no need for a separate widget. The ImageCropper on its own is sufficient (using cropOnTap=true). However there will be case when you want to trigger it manually (e.g. wrap the ImageCropper and overlay the Button in a Stack).
```
- ImageCropper:
    id: myCropper
    source: ...
    styles: ...
    options: ...                      # all the crop options go here
    cropOnTap: true|false    # default TRUE to initiate the crop on tap, FALSE to trigger it manually
    onCropped:                    # initiate an action, e.g. server call to update the cropped image
    
- Button:
     label: Crop Image
     onTap: |
       myCropper.crop();
```

**Comment by vinothvino42** [comment]
@vusters Can you check this EDL?

```yaml
View:
  header:
    title: Image Cropper
  styles:
    scrollableView: true
  body:
    Column:
      styles:
        gap: 24
        padding: 24
      children:
        - ImageCropper:
            id: imageCropperId
           source: example.png
            styles:
              {
                cornerDotColor: green,
                maskColor: blue,
                cornerRadius: 25,
                height: 120,
                width: 100,
                borderWidth: 1,
                fit: cover,
              }
            onCropTap: true
            onCropped:
              executeCode:
                body: |
                  //@code
                  console.log("Image Cropped Successfully");

        - Button:
            label: Crop Image
            onTap:
              executeCode:
                body: |
                  //@code
                  imageCropperId.crop();
```

So after cropping the image, I need to set the cropped image to the ImageCropper itself? No need to expose a file path like `ensemble.data.file` to the user?

**Comment by vusters** [comment]
> @vusters Can you check this EDL? I updated it.
> 
> ```yaml
> View:
>   header:
>     title: Image Cropper
>   styles:
>     scrollableView: true
>   body:
>     Column:
>       styles:
>         gap: 24
>         padding: 24
>       children:
>         - ImageCropper:
>             id: imageCropperId
>            source: example.png
>             styles:
>               {
>                 cornerDotColor: green,
>                 maskColor: blue,
>                 cornerRadius: 25,
>                 height: 120,
>                 width: 100,
>                 borderWidth: 1,
>                 fit: cover,
>               }
>             onCropTap: true
>             onCropped:
>               executeCode:
>                 body: |
>                   //@code
>                   console.log("Image Cropped Successfully");
> 
>         - Button:
>             label: Crop Image
>             onTap:
>               executeCode:
>                 body: |
>                   //@code
>                   imageCropperId.crop();
> ```
> 
> So after cropping the image, I need to set the cropped image to the ImageCropper itself? No need to expose a file path like `ensemble.data.file` to the user?
Yes, the ImageCropper's source will change to the cropped image. However on the action `onCropped`, the user may need to use File upload to upload the newly cropped image. In that case there has to be a way for the user to access the file. See how FileUpload is used to make it consistent there

**Comment by vusters** [comment]
Make sure you use `cropOnTap` name or something that makes sense. We are saying we will initiate the cropping when the user tap on the widget. What you had there "onCropTap" does not make sense.

**Review by vusters — APPROVED** [review]

**Comment by kmahmood74** [comment]
@vinothvino42 this was approved last week. Please merge and close.

File: lib/util/utils.dart [hunk 1]
```diff
@@ -760,4 +760,9 @@ class Utils {
       return dataContext.containsKey(key) ? dataContext[key]! : match.group(0)!;
     });
   }
+
+  static String? capitalizeFirstLetter(String? value) {
+    if (value == null || value.isEmpty) return value;
+    return '${value[0].toUpperCase()}${value.substring(1).toLowerCase()}';
```
[the remark this answers is not in the document]
Comment (vinothvino42) [reply lib/util/utils.dart:766]: [CustomCropShape](https://pub.dev/documentation/custom_image_crop/latest/custom_image_crop/CustomCropShape.html)
All these enum values start with the first letter as an uppercase one. So I am just capitalizing only the first character and others will be in lowercase.

File: lib/widget/image_cropper.dart [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+  ImageProvider buildImageProvider(String source, BoxFit? fit) {
+    if (isSvg()) {
+      return buildSvgImageProvider(source, fit);
+    }
+    return buildNonSvgImageProvider(source, fit);
+  }
+
+  ImageProvider buildNonSvgImageProvider(String source, BoxFit? fit) {
+    if (source.startsWith('https://') || source.startsWith('http://')) {
+      return CachedNetworkImageProvider(
```
[the remark this answers is not in the document]
Comment (vinothvino42) [reply lib/widget/image_cropper.dart]: I changed it to `Image.network`. Can you check again?

File: lib/framework/data_context.dart [hunk 3]
```diff
@@ -846,6 +846,29 @@ class ModifiableAPIResponse extends APIResponse {
   }
 }
 
+class CroppedFileData with Invokable {
```
[the remark this answers is not in the document]
Comment (vinothvino42) [reply lib/framework/data_context.dart]: Removed it. Used it previously

File: lib/framework/data_context.dart [hunk 4]
```diff
@@ -879,6 +902,13 @@ class File {
         path = file['path'],
         bytes = file['bytes'];
 
+  File.fromCropImage(Map<String, dynamic> file)
```
[the remark this answers is not in the document]
Comment (vinothvino42) [reply lib/framework/data_context.dart]: Removed it. Used it previously

File: lib/util/utils.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (evshi) [reply lib/util/utils.dart:766]: This doesn't seem like a generic utility function then and should just move it to where it's being used.

File: pubspec.yaml [hunk 5]
```diff
@@ -100,6 +101,9 @@ dependencies:
   dart_jsonwebtoken: ^2.8.2
   flutter_dotenv: ^5.1.0
   get_it: ^7.6.0
+  google_sign_in: ^6.1.4
+  sign_in_with_apple: ^5.0.0
```
[the remark this answers is not in the document]
Comment (evshi) [reply pubspec.yaml:105]: @vinothvino42 please remove these

File: lib/util/utils.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (vinothvino42) [reply lib/util/utils.dart]: Done

File: pubspec.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (vinothvino42) [reply pubspec.yaml]: Removed it. Can you review the PR again and let me know whether I can merge this PR

File: lib/widget/image_cropper.dart [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+  Color? strokeColor;
+  bool isRotate = true;
+  bool isMove = true;
+  bool isScale = true;
+  EnsembleAction? onCropped;
+  CustomImageCropController cropController = CustomImageCropController();
+}
+
+class EnsembleImageCropperState extends WidgetState<EnsembleImageCropper>
+    with CropAction {
```
Comment (evshi) [inline lib/widget/image_cropper.dart:125]: I don't understand this relationship - why does the state have all the logic? I think all the logic should be implemented on the controller and the state just fields in the values. @vusters WDYT?

File: lib/widget/image_cropper.dart [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+          _controller.isMove = Utils.getBool(value, fallback: true),
+      'isScale': (value) =>
+          _controller.isScale = Utils.getBool(value, fallback: true),
+      'onCropped': (funcDefinition) => _controller.onCropped =
+          EnsembleAction.fromYaml(funcDefinition, initiator: this)
+    };
+  }
+}
+
+mixin CropAction on WidgetState<EnsembleImageCropper> {
```
Comment (evshi) [inline lib/widget/image_cropper.dart:96]: Why does this mixin exist?

File: lib/widget/image_cropper.dart [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+  void rotateLeft() {
+    widget.controller.cropController
+        .addTransition(CropImageData(angle: -pi / 4));
+  }
+
+  @override
+  void rotateRight() {
+    widget.controller.cropController
+        .addTransition(CropImageData(angle: pi / 4));
+  }
```
Comment (evshi) [inline lib/widget/image_cropper.dart:186]: Why are these just not implemented on the controller directly?

Comment (vinothvino42) [reply lib/widget/image_cropper.dart]: I moved crop and rotate, reset related methods to the controller. Can you review it again?

File: lib/widget/image_cropper.dart (hunk 6 above)
Comment (evshi) [reply lib/widget/image_cropper.dart:148]: Can't the context be set on the controller during the build method? In any case, this seems good enough for now.

**Review by evshi — APPROVED** [review]

**Pull request merged**